internal/consensus/reactor.go

```diff
@@ -229,34 +229,43 @@ func (r *Reactor) OnStop() {
 		r.state.Wait()
 	}
 
-	r.mtx.Lock()
-	peers := r.peers
-	r.mtx.Unlock()
-
-	// wait for all spawned peer goroutines to gracefully exit
-	for _, ps := range peers {
-		ps.closer.Close()
-	}
-	for _, ps := range peers {
-		ps.broadcastWG.Wait()
-	}
+	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
+	// p2p Channels should execute Close().
+	close(r.closeCh)
 
 	// Close the StateChannel goroutine separately since it uses its own channel
 	// to signal closure.
 	close(r.stateCloseCh)
 	<-r.stateCh.Done()
 
-	// Close closeCh to signal to all spawned goroutines to gracefully exit. All
-	// p2p Channels should execute Close().
-	close(r.closeCh)
-
 	// Wait for all p2p Channels to be closed before returning. This ensures we
 	// can easily reason about synchronization of all p2p Channels and ensure no
 	// panics will occur.
 	<-r.voteSetBitsCh.Done()
 	<-r.dataCh.Done()
 	<-r.voteCh.Done()
 	<-r.peerUpdates.Done()
+
+	peers := make(map[types.NodeID]*PeerState)
+	r.mtx.Lock()
+
+	// Here, we make a copy of the map of peer states.
+	// The code below proceeds to access entries from the map outside of a lock.
+	// Goroutines launched by other functions may be deleting from the map but should
+	// no longer be appending to it since we waited for the peerUpdates channel to complete.
+	// Each of the methods below is safe to perform concurrently on the peers.
+	for id, state := range r.peers {
+		peers[id] = state
+	}
+	r.mtx.Unlock()
+
+	// wait for all spawned peer goroutines to gracefully exit
+	for _, ps := range peers {
+		ps.closer.Close()
+	}
+	for _, ps := range peers {
+		ps.broadcastWG.Wait()
```

Comment:
do we expect these to take a while or is it just belt and suspenders? Seems like we could wait in parallel, but maybe that's overkill.

Reply:
I think it doesn't really matter. We'll just end up waiting for the slowest `Wait` and no longer either way.

Reply:
Well, we might wait longer if the first one in the collection is the slowest—but I do not think we care, since this is an OnStop handler anyway.

Reply:
I suppose, but the general idea is that the atomics on the other side of the `Wait` will start to be flipped by goroutines after we call the above `Close` method and we're just hanging out and checking to see if they've been flipped yet. We could check each atomic in a separate thread but ultimately the difference seems negligible.

Reply:
Oh, I agree—I don't think it's important to try to optimize it at all, especially on a shutdown path.